File: languages/php/XANDEncrypt.php
<?php

class XANDEncrypt {
    private $basePassword;
    private $keySize = 32; // 256 bits
    private $nonceSize = 16;
    private $paddingSize = 16;

    public function __construct($password) {
        $this->basePassword = $password;
    }

    /**
     * Generate a hash from the password
     */
    private function hashPassword($password) {
        return hash('sha256', $password, true);
    }

    /**
     * Generate HMAC
     */
    private function generateHMAC($key, $data) {
        return hash_hmac('sha256', $data, $key, true);
    }

    /**
     * Generate time-shifted keys based on current timestamp
     */
    private function generateTimeShiftedKeys($baseKey, $timestamp) {
        $timeBuffer = pack('J', $timestamp); // 8-byte timestamp
        
        // Create three different keys for the three layers
        $key1 = $this->generateHMAC($baseKey, $timeBuffer);
        $key2 = $this->generateHMAC($key1, 'layer2');
        $key3 = $this->generateHMAC($key2, 'layer3');
        
        return [
            'key1' => $key1,
            'key2' => $key2,
            'key3' => $key3,
            'timestamp' => $timestamp
        ];
    }

    /**
     * Generate random nonce and derive padding
     */
    private function generateNonceAndPadding() {
        $nonce = random_bytes($this->nonceSize);
        $paddingLength = ord($nonce[0]) % $this->paddingSize + 1;
        $padding = random_bytes($paddingLength);
        
        return [
            'nonce' => $nonce,
            'padding' => $padding
        ];
    }

    /**
     * Layer 1: Bitwise scrambling with time-shifted key
     */
    private function layer1Scramble($data, $key) {
        $result = '';
        $dataLength = strlen($data);
        $keyLength = strlen($key);
        
        for ($i = 0; $i < $dataLength; $i++) {
            $keyByte = ord($key[$i % $keyLength]);
            $dataByte = ord($data[$i]);
            
            // Complex bitwise operations
            $scrambled = $dataByte ^ $keyByte;
            $scrambled = (($scrambled << 3) | ($scrambled >> 5)) & 0xFF; // Rotate left by 3
            $scrambled = $scrambled ^ ((($keyByte << 1) | ($keyByte >> 7)) & 0xFF);
            
            $result .= chr($scrambled);
        }
        
        return $result;
    }

    /**
     * Layer 2: Advanced bit manipulation
     */
    private function layer2Manipulate($data, $key) {
        $result = '';
        $dataLength = strlen($data);
        $keyLength = strlen($key);
        
        for ($i = 0; $i < $dataLength; $i++) {
            $keyByte = ord($key[$i % $keyLength]);
            $dataByte = ord($data[$i]);
            
            // Position-based bit rotation
            $rotation = ($i + $keyByte) % 8;
            $manipulated = $dataByte;
            
            if ($rotation > 0) {
                $manipulated = (($manipulated << $rotation) | ($manipulated >> (8 - $rotation))) & 0xFF;
            }
            
            // XOR with position-dependent key
            $manipulated = $manipulated ^ (($keyByte + $i) & 0xFF);
            
            $result .= chr($manipulated);
        }
        
        return $result;
    }

    /**
     * Layer 3: AES-256 encryption
     */
    private function layer3Encrypt($data, $key) {
        $iv = random_bytes(16);
        
        // Use OpenSSL for AES encryption
        $encrypted = openssl_encrypt(
            $data,
            'AES-256-CBC',
            $key,
            OPENSSL_RAW_DATA,
            $iv
        );
        
        return [
            'encrypted' => $encrypted,
            'iv' => $iv
        ];
    }

    /**
     * Layer 3: AES-256 decryption
     */
    private function layer3Decrypt($data, $key, $iv) {
        return openssl_decrypt(
            $data,
            'AES-256-CBC',
            $key,
            OPENSSL_RAW_DATA,
            $iv
        );
    }

    /**
     * Apply key decay - modify keys after use
     */
    private function applyKeyDecay($keys) {
        $decayedKeys = [];
        
        foreach ($keys as $keyName => $key) {
            if ($keyName === 'timestamp') {
                $decayedKeys[$keyName] = $key;
                continue;
            }
            
            // Apply decay by XORing with a derived value
            $timeStr = (string)(microtime(true) * 1000);
            $decayValue = hash('sha256', $key . $timeStr, true);
            
            $decayedKey = '';
            $keyLength = strlen($key);
            $decayLength = strlen($decayValue);
            
            for ($i = 0; $i < $keyLength; $i++) {
                $decayedKey .= chr(ord($key[$i]) ^ ord($decayValue[$i % $decayLength]));
            }
            
            $decayedKeys[$keyName] = $decayedKey;
        }
        
        return $decayedKeys;
    }

    /**
     * Encrypt data using triple-layer encryption
     */
    public function encrypt($data) {
        // Generate base key from password
        $baseKey = $this->hashPassword($this->basePassword);
        
        // Generate time-shifted keys
        $timestamp = (int)(microtime(true) * 1000);
        $keys = $this->generateTimeShiftedKeys($baseKey, $timestamp);
        
        // Generate nonce and padding
        $nonceAndPadding = $this->generateNonceAndPadding();
        $nonce = $nonceAndPadding['nonce'];
        $padding = $nonceAndPadding['padding'];
        
        // Add padding to data
        $paddedData = $data . $padding;
        
        // Layer 1: Bitwise scrambling
        $layer1Result = $this->layer1Scramble($paddedData, $keys['key1']);
        
        // Layer 2: Advanced bit manipulation
        $layer2Result = $this->layer2Manipulate($layer1Result, $keys['key2']);
        
        // Layer 3: AES encryption
        $layer3Result = $this->layer3Encrypt($layer2Result, $keys['key3']);
        
        // Apply key decay
        $this->applyKeyDecay($keys);
        
        // Combine all components
        $result = [
            'encrypted' => base64_encode($layer3Result['encrypted']),
            'iv' => base64_encode($layer3Result['iv']),
            'nonce' => base64_encode($nonce),
            'timestamp' => $timestamp,
            'padding_length' => strlen($padding)
        ];
        
        return json_encode($result);
    }

    /**
     * Decrypt data using triple-layer decryption
     */
    public function decrypt($encryptedData) {
        $data = json_decode($encryptedData, true);
        
        // Parse components
        $encrypted = base64_decode($data['encrypted']);
        $iv = base64_decode($data['iv']);
        $nonce = base64_decode($data['nonce']);
        $timestamp = $data['timestamp'];
        $paddingLength = $data['padding_length'];
        
        // Generate base key from password
        $baseKey = $this->hashPassword($this->basePassword);
        
        // Regenerate time-shifted keys (must use same timestamp)
        $keys = $this->generateTimeShiftedKeys($baseKey, $timestamp);
        
        // Layer 3: AES decryption
        $layer3Result = $this->layer3Decrypt($encrypted, $keys['key3'], $iv);
        
        // Layer 2: Reverse bit manipulation
        $layer2Result = $this->layer2Manipulate($layer3Result, $keys['key2']);
        
        // Layer 1: Reverse bitwise scrambling
        $layer1Result = $this->layer1Scramble($layer2Result, $keys['key1']);
        
        // Remove padding
        $originalData = substr($layer1Result, 0, -$paddingLength);
        
        return $originalData;
    }

    /**
     * Generate a new key pair for asymmetric operations (if needed)
     */
    public function generateKeyPair() {
        $privateKey = random_bytes(32);
        $publicKey = hash('sha256', base64_encode($privateKey), true);
        
        return [
            'private_key' => base64_encode($privateKey),
            'public_key' => base64_encode($publicKey)
        ];
    }
} 

File: languages/java/XANDEncrypt.java
import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import org.json.JSONObject;

public class XANDEncrypt {
    private final String basePassword;
    private final int keySize = 32; // 256 bits
    private final int nonceSize = 16;
    private final int paddingSize = 16;
    private final SecureRandom secureRandom = new SecureRandom();

    public XANDEncrypt(String password) {
        this.basePassword = password;
    }

    /**
     * Convert password to SHA-256 hash using Java MessageDigest
     * JCE (Java Cryptography Extension) provides optimized cryptographic operations
     * StandardCharsets.UTF_8 ensures consistent encoding across different platforms
     */
    private byte[] hashPassword(String password) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        return digest.digest(password.getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Generate HMAC for key derivation
     */
    private byte[] generateHMAC(byte[] key, byte[] data) throws Exception {
        javax.crypto.Mac mac = javax.crypto.Mac.getInstance("HmacSHA256");
        javax.crypto.spec.SecretKeySpec secretKeySpec = new javax.crypto.spec.SecretKeySpec(key, "HmacSHA256");
        mac.init(secretKeySpec);
        return mac.doFinal(data);
    }

    /**
     * Generate three time-dependent keys using HMAC-SHA256
     * Manual byte extraction from long ensures big-endian byte order
     * Java lacks built-in long-to-byte conversion, requiring manual bit manipulation
     */
    private Map<String, Object> generateTimeShiftedKeys(byte[] baseKey, long timestamp) throws Exception {
        byte[] timeBuffer = new byte[8];
        for (int i = 0; i < 8; i++) {
            timeBuffer[i] = (byte) (timestamp >> (56 - i * 8));
        }

        // Each layer gets its own unique key derived from the previous one
        // Java's Mac class is thread-safe and uses native crypto when available
        byte[] key1 = generateHMAC(baseKey, timeBuffer);
        byte[] key2 = generateHMAC(key1, "layer2".getBytes(StandardCharsets.UTF_8));
        byte[] key3 = generateHMAC(key2, "layer3".getBytes(StandardCharsets.UTF_8));

        Map<String, Object> keys = new HashMap<>();
        keys.put("key1", key1);
        keys.put("key2", key2);
        keys.put("key3", key3);
        keys.put("timestamp", timestamp);
        return keys;
    }

    /**
     * Generate cryptographically secure random nonce and padding
     * SecureRandom uses OS random number generator for cryptographic security
     * & 0xFF converts signed Java bytes to unsigned values (0-255 range)
     */
    private Map<String, byte[]> generateNonceAndPadding() {
        byte[] nonce = new byte[nonceSize];
        secureRandom.nextBytes(nonce);
        
        int paddingLength = (nonce[0] & 0xFF) % paddingSize + 1;
        byte[] padding = new byte[paddingLength];
        secureRandom.nextBytes(padding);
        
        Map<String, byte[]> result = new HashMap<>();
        result.put("nonce", nonce);
        result.put("padding", padding);
        return result;
    }

    /**
     * First encryption layer: XOR and bit rotation operations
     * Java bytes are signed (-128 to 127), requiring int conversion for bitwise ops
     * >>> (unsigned right shift) prevents sign extension during bit manipulation
     */
    private byte[] layer1Scramble(byte[] data, byte[] key) {
        byte[] result = new byte[data.length];
        
        for (int i = 0; i < data.length; i++) {
            byte keyByte = key[i % key.length];
            byte dataByte = data[i];
            
            // Do some bit magic to scramble the data
            // Java bitwise ops work on int, so we cast bytes to int first
            int scrambled = dataByte ^ keyByte;
            scrambled = ((scrambled << 3) | (scrambled >>> 5)) & 0xFF; // Rotate left by 3
            scrambled = scrambled ^ (((keyByte << 1) | (keyByte >>> 7)) & 0xFF);
            
            result[i] = (byte) scrambled;
        }
        
        return result;
    }

    /**
     * Second encryption layer: Position-dependent bit rotation
     * Rotation amount calculated from byte position and key value
     * Conditional rotation prevents unnecessary operations when rotation is zero
     */
    private byte[] layer2Manipulate(byte[] data, byte[] key) {
        byte[] result = new byte[data.length];
        
        for (int i = 0; i < data.length; i++) {
            byte keyByte = key[i % key.length];
            byte dataByte = data[i];
            
            // Rotate bits based on position and key
            int rotation = (i + (keyByte & 0xFF)) % 8;
            int manipulated = dataByte & 0xFF;
            
            if (rotation > 0) {
                manipulated = ((manipulated << rotation) | (manipulated >>> (8 - rotation))) & 0xFF;
            }
            
            // Mix in some position-dependent randomness
            manipulated = manipulated ^ ((keyByte + i) & 0xFF);
            
            result[i] = (byte) manipulated;
        }
        
        return result;
    }

    /**
     * Third encryption layer: AES-256-CBC encryption
     * Uses Java Cipher class with PKCS5Padding for standard compliance
     * CBC mode requires unique IV for each encryption operation
     */
    private Map<String, byte[]> layer3Encrypt(byte[] data, byte[] key) throws Exception {
        byte[] iv = new byte[16];
        secureRandom.nextBytes(iv);
        
        SecretKeySpec secretKeySpec = new SecretKeySpec(key, "AES");
        IvParameterSpec ivSpec = new IvParameterSpec(iv);
        
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec, ivSpec);
        
        byte[] encrypted = cipher.doFinal(data);
        
        Map<String, byte[]> result = new HashMap<>();
        result.put("encrypted", encrypted);
        result.put("iv", iv);
        return result;
    }

    /**
     * Layer 3: AES-256 decryption
     */
    private byte[] layer3Decrypt(byte[] data, byte[] key, byte[] iv) throws Exception {
        SecretKeySpec secretKeySpec = new SecretKeySpec(key, "AES");
        IvParameterSpec ivSpec = new IvParameterSpec(iv);
        
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.DECRYPT_MODE, secretKeySpec, ivSpec);
        
        return cipher.doFinal(data);
    }

    /**
     * Apply key decay by XORing with time-based hash
     * Prevents key reuse attacks by modifying keys after each operation
     * Uses current timestamp to ensure uniqueness of decay value
     */
    private Map<String, Object> applyKeyDecay(Map<String, Object> keys) throws Exception {
        Map<String, Object> decayedKeys = new HashMap<>();
        
        for (Map.Entry<String, Object> entry : keys.entrySet()) {
            String keyName = entry.getKey();
            Object key = entry.getValue();
            
            if (keyName.equals("timestamp")) {
                decayedKeys.put(keyName, key);
                continue;
            }
            
            if (key instanceof byte[]) {
                byte[] keyBytes = (byte[]) key;
                // Mix in some fresh randomness to age the key
                String timeStr = String.valueOf(System.currentTimeMillis());
                byte[] decayValue = hashPassword(new String(keyBytes, StandardCharsets.UTF_8) + timeStr);
                
                byte[] decayedKey = new byte[keyBytes.length];
                for (int i = 0; i < keyBytes.length; i++) {
                    decayedKey[i] = (byte) (keyBytes[i] ^ decayValue[i % decayValue.length]);
                }
                
                decayedKeys.put(keyName, decayedKey);
            } else {
                decayedKeys.put(keyName, key);
            }
        }
        
        return decayedKeys;
    }

    /**
     * Main encryption function implementing triple-layer encryption
     * Processes data through bit scrambling, position-dependent rotation, and AES-256
     * Returns JSON string containing encrypted data and metadata
     */
    public String encrypt(String data) throws Exception {
        byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
        
        // Generate base key from password
        byte[] baseKey = hashPassword(basePassword);
        
        // Generate time-shifted keys
        long timestamp = System.currentTimeMillis();
        Map<String, Object> keys = generateTimeShiftedKeys(baseKey, timestamp);
        
        // Generate nonce and padding
        Map<String, byte[]> nonceAndPadding = generateNonceAndPadding();
        byte[] nonce = nonceAndPadding.get("nonce");
        byte[] padding = nonceAndPadding.get("padding");
        
        // Add padding to data
        byte[] paddedData = new byte[dataBytes.length + padding.length];
        System.arraycopy(dataBytes, 0, paddedData, 0, dataBytes.length);
        System.arraycopy(padding, 0, paddedData, dataBytes.length, padding.length);
        
        // Layer 1: Bitwise scrambling
        byte[] layer1Result = layer1Scramble(paddedData, (byte[]) keys.get("key1"));
        
        // Layer 2: Advanced bit manipulation
        byte[] layer2Result = layer2Manipulate(layer1Result, (byte[]) keys.get("key2"));
        
        // Layer 3: AES encryption
        Map<String, byte[]> layer3Result = layer3Encrypt(layer2Result, (byte[]) keys.get("key3"));
        
        // Apply key decay
        applyKeyDecay(keys);
        
        // Combine all components
        JSONObject result = new JSONObject();
        result.put("encrypted", Base64.getEncoder().encodeToString(layer3Result.get("encrypted")));
        result.put("iv", Base64.getEncoder().encodeToString(layer3Result.get("iv")));
        result.put("nonce", Base64.getEncoder().encodeToString(nonce));
        result.put("timestamp", timestamp);
        result.put("paddingLength", padding.length);
        
        return result.toString();
    }

    /**
     * Decrypt data using triple-layer decryption
     */
    public String decrypt(String encryptedData) throws Exception {
        JSONObject data = new JSONObject(encryptedData);
        
        // Parse components
        byte[] encrypted = Base64.getDecoder().decode(data.getString("encrypted"));
        byte[] iv = Base64.getDecoder().decode(data.getString("iv"));
        byte[] nonce = Base64.getDecoder().decode(data.getString("nonce"));
        long timestamp = data.getLong("timestamp");
        int paddingLength = data.getInt("paddingLength");
        
        // Generate base key from password
        byte[] baseKey = hashPassword(basePassword);
        
        // Regenerate time-shifted keys (must use same timestamp)
        Map<String, Object> keys = generateTimeShiftedKeys(baseKey, timestamp);
        
        // Layer 3: AES decryption
        byte[] layer3Result = layer3Decrypt(encrypted, (byte[]) keys.get("key3"), iv);
        
        // Layer 2: Reverse bit manipulation
        byte[] layer2Result = layer2Manipulate(layer3Result, (byte[]) keys.get("key2"));
        
        // Layer 1: Reverse bitwise scrambling
        byte[] layer1Result = layer1Scramble(layer2Result, (byte[]) keys.get("key1"));
        
        // Remove padding
        byte[] originalData = new byte[layer1Result.length - paddingLength];
        System.arraycopy(layer1Result, 0, originalData, 0, originalData.length);
        
        return new String(originalData, StandardCharsets.UTF_8);
    }

    /**
     * Generate a new key pair for asymmetric operations (if needed)
     */
    public Map<String, String> generateKeyPair() throws Exception {
        KeyGenerator keyGen = KeyGenerator.getInstance("AES");
        keyGen.init(256);
        SecretKey secretKey = keyGen.generateKey();
        
        byte[] publicKey = hashPassword(Base64.getEncoder().encodeToString(secretKey.getEncoded()));
        
        Map<String, String> keyPair = new HashMap<>();
        keyPair.put("private_key", Base64.getEncoder().encodeToString(secretKey.getEncoded()));
        keyPair.put("public_key", Base64.getEncoder().encodeToString(publicKey));
        
        return keyPair;
    }
} 